#################
#### imports ####
#################
import pandas as pd

##################
#### SETTINGS ####
##################

##### check snakemake min version #####
from snakemake.utils import min_version

min_version("7.15.1")


##### report workflow #####
report: "report/workflow.rst"


##### General wirldcard constrains #####
wildcard_constraints:
    project="[^/\.]+",
    assignment="[^/\.]+",
    condition="[^/_\.]+",
    replicate="[^/_\.]+",
    type="(DNA)|(RNA)",
    config="[^/\.]+",


##### localrules #####
localrules:
    all,
    counts_demultiplex_aggregate,


# common workflow
include: "rules/common.smk"


###################
#### ALL RULES ####
###################


rule all:
    input:
        ## assignments
        getAssignmentConfig_helper(
            "results/assignment/{assignment}/assignment_barcodes.{config}.sorted.tsv.gz",
        ),
        getAssignmentConfig_helper(
            "results/assignment/{assignment}/statistic/assigned_counts.{config}.tsv.gz",
        ),
        getAssignmentConfig_helper(
            "results/assignment/{assignment}/statistic/assignment.{config}.png",
        ),
        getAssignment_helper(
            "results/assignment/{assignment}/statistic/total_counts.tsv.gz"
        ),
        getAssignment_helper(
            "results/assignment/{assignment}/statistic/assignment/bam_stats.txt",
        ),
        ## experiments
        # statistic BC nucleotide composition
        getOutputProjectConditionReplicateType_helper(
            "results/experiments/{project}/statistic/counts/BCNucleotideComposition.{condition}_{replicate}_{type}.tsv.gz"
        ),
        # counts statistic
        expand(
            "results/experiments/{project}/statistic/counts.{countType}.tsv",
            project=getProjects(),
            countType=["raw", "filtered"],
        ),
        getOutputProjectConditionReplicateType_helper(
            "results/experiments/{project}/statistic/counts.freqUMIs.{condition}_{replicate}_{type}.txt"
        ),
        getOutputProjectConfig_helper(
            "results/experiments/{project}/statistic/statistic_bc_correlation_merged_{config}.tsv",
            True,
        ),
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/statistic_assigned_bc_correlation_merged_{assignment}_{config}.tsv",
            True,
        ),
        # bc_overlap statistic
        getOutputProjectConfig_helper(
            "results/experiments/{project}/statistic/bc_overlap.counts.{config}.tsv",
            True,
        ),
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/bc_overlap.assigned_counts.{config}.{assignment}.tsv",
            True,
        ),
        # assignment
        getOutputProjectConditionAssignmentConfig_helper(
            "results/experiments/{project}/assigned_counts/{assignment}/{config}/{condition}_allreps_merged.tsv.gz"
        ),
        # assignment statistic
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/statistic_assigned_counts_merged_{assignment}_{config}.tsv"
        ),
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/statistic_oligo_correlation_merged_{assignment}_{config}.tsv",
            True,
        ),
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/statistic_assigned_counts_single_{assignment}_{config}.tsv",
            True,
        ),
        # variants
        getOutputVariants_helper(
            "results/experiments/{project}/statistic/variants/{assignment}/{config}/correlation_variantTable.tsv",
            True,
        ),
        getOutputVariants_helper(
            "results/experiments/{project}/variants/{assignment}/{config}/{condition}_variantTable.tsv.gz",
            False,
        ),


rule all_assignments:
    input:
        getAssignmentConfig_helper(
            "results/assignment/{assignment}/assignment_barcodes.{config}.sorted.tsv.gz",
        ),
        getAssignmentConfig_helper(
            "results/assignment/{assignment}/statistic/assigned_counts.{config}.tsv.gz",
        ),
        getAssignmentConfig_helper(
            "results/assignment/{assignment}/statistic/assignment.{config}.png",
        ),
        getAssignment_helper(
            "results/assignment/{assignment}/statistic/total_counts.tsv.gz"
        ),
        getAssignment_helper(
            "results/assignment/{assignment}/statistic/assignment/bam_stats.txt",
        ),


rule all_experiments:
    input:
        # stats BC nucleotide composition
        getOutputProjectConditionReplicateType_helper(
            "results/experiments/{project}/statistic/counts/BCNucleotideComposition.{condition}_{replicate}_{type}.tsv.gz"
        ),
        # counts
        getOutputProjectConditionReplicateType_helper(
            "results/experiments/{project}/counts/{condition}_{replicate}_{type}.bam",
            skip={"demultiplex": True},
        ),
        getOutputProjectConditionReplicateType_helper(
            "results/experiments/{project}/counts/{condition}_{replicate}_{type}_final_counts.tsv.gz"
        ),
        # counts statistic
        expand(
            "results/experiments/{project}/statistic/counts.{countType}.tsv",
            project=getProjects(),
            countType=["raw", "filtered"],
        ),
        getOutputProjectConditionReplicateType_helper(
            "results/experiments/{project}/statistic/counts.freqUMIs.{condition}_{replicate}_{type}.txt"
        ),
        getOutputProjectConfig_helper(
            "results/experiments/{project}/statistic/statistic_bc_correlation_merged_{config}.tsv",
            True,
        ),
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/statistic_assigned_bc_correlation_merged_{assignment}_{config}.tsv",
            True,
        ),
        # bc_overlap statistic
        getOutputProjectConfig_helper(
            "results/experiments/{project}/statistic/bc_overlap.counts.{config}.tsv",
            True,
        ),
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/bc_overlap.assigned_counts.{config}.{assignment}.tsv",
            True,
        ),
        # assignment
        getOutputProjectConditionAssignmentConfig_helper(
            "results/experiments/{project}/assigned_counts/{assignment}/{config}/{condition}_allreps_merged.tsv.gz"
        ),
        # assignment statistic
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/statistic_assigned_counts_merged_{assignment}_{config}.tsv"
        ),
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/statistic_oligo_correlation_merged_{assignment}_{config}.tsv",
            True,
        ),
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/statistic_assigned_counts_single_{assignment}_{config}.tsv",
            True,
        ),
        # variants
        getOutputVariants_helper(
            "results/experiments/{project}/statistic/variants/{assignment}/{config}/correlation_variantTable.tsv",
            True,
        ),


rule all_experiments_counts_stats:
    input:
        getOutputProjectConditionReplicateType_helper(
            "results/experiments/{project}/statistic/counts.freqUMIs.{condition}_{replicate}_{type}.txt"
        ),
        getOutputProjectConfig_helper(
            "results/experiments/{project}/statistic/statistic_bc_correlation_merged_{config}.tsv",
            True,
        ),


rule all_bc_overlap_statistic:
    input:
        getOutputProjectConfig_helper(
            "results/experiments/{project}/statistic/bc_overlap.counts.{config}.tsv",
            True,
        ),
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/bc_overlap.assigned_counts.{config}.{assignment}.tsv",
            True,
        ),


rule all_experiments_assignments:
    input:
        getOutputProjectConditionAssignmentConfig_helper(
            "results/experiments/{project}/assigned_counts/{assignment}/{config}/{condition}_allreps_merged.tsv.gz"
        ),


rule all_experiments_assignments_statistic:
    input:
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/statistic_assigned_counts_merged_{assignment}_{config}.tsv"
        ),
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/statistic_oligo_correlation_merged_{assignment}_{config}.tsv",
            True,
        ),
        getOutputProjectAssignmentConfig_helper(
            "results/experiments/{project}/statistic/statistic_assigned_counts_single_{assignment}_{config}.tsv",
            True,
        ),


rule all_stats_BCNucleotideComposition:
    input:
        getOutputProjectConditionReplicateType_helper(
            "results/experiments/{project}/statistic/counts/BCNucleotideComposition.{condition}_{replicate}_{type}.tsv.gz"
        ),


rule all_variants:
    input:
        getOutputVariants_helper(
            "results/experiments/{project}/statistic/variants/{assignment}/{config}/correlation_variantTable.tsv",
            True,
        ),
        getOutputVariants_helper(
            "results/experiments/{project}/variants/{assignment}/{config}/{condition}_variantTable.tsv.gz",
            False,
        ),


###################
## SUB-WORKFLOWS ##
###################


# assignment workflow
include: "rules/assignment.smk"
# count workflow
include: "rules/counts.smk"
# assignimg BCs to oligos
include: "rules/assigned_counts.smk"
# all statistics
include: "rules/statistic.smk"
# variant subworkflow (creating variant effects)
include: "rules/variants.smk"
